package/src/lib.rs

```diff
@@ -62,23 +62,76 @@ impl GitDependency {
     }
 }
 
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
+/// A dependency that comes from the global package index.
+pub struct IndexDependency {
+    pub id: index::Id,
+    pub version: VersionReq,
+}
+
 /// A source includes the place to fetch a package from (e.g. git or a registry),
 /// along with possibly some narrowing-down of the allowed versions (e.g. a range
 /// of versions, or a git commit id).
 #[derive(Clone, Debug, PartialEq, Eq, Hash)]
 pub enum Dependency {
     Git(GitDependency),
     Path(PathBuf),
+    Index(IndexDependency),
 }
 
 impl Dependency {
-    pub fn matches(&self, entry: &LockFileDep) -> bool {
+    pub fn matches(&self, entry: &LockFileDep, precise: &LockPrecisePkg) -> bool {
         match self {
             Dependency::Git(git) => match &entry.spec {
                 Some(locked_git) => git == locked_git,
                 None => false,
             },
-            Dependency::Path { .. } => true,
+            Dependency::Path(_) => true,
+            Dependency::Index(i) => {
+                if let LockPrecisePkg::Index { id, version } = precise {
+                    i.id == *id && i.version.matches(version)
+                } else {
+                    false
+                }
+            }
+        }
+    }
+
+    pub fn as_index_dep(self, parent_id: &index::Id) -> Result<IndexDependency, Error> {
+        match self {
+            Dependency::Index(i) => Ok(i),
+            Dependency::Git(g) => Err(Error::InvalidIndexDep {
+                id: parent_id.clone(),
+                dep: Box::new(crate::UnversionedDependency::Git(g)),
+            }),
+            Dependency::Path(path) => Err(Error::InvalidIndexDep {
+                id: parent_id.clone(),
+                dep: Box::new(crate::UnversionedDependency::Path(path)),
+            }),
+        }
+    }
+
+    pub fn as_unversioned(self) -> Option<UnversionedDependency> {
+        match self {
+            Dependency::Index(_i) => None,
+            Dependency::Git(g) => Some(UnversionedDependency::Git(g)),
+            Dependency::Path(p) => Some(UnversionedDependency::Path(p)),
+        }
+    }
+}
+
+/// The same as [`Dependency`], but only for the packages that have fixed, unresolvable, versions.
```

Comment:
What does "unresolvable" means in this context? Just like we don't get to pick a different one, like rigid version constraints, or something like a precise git ref?

Reply:
It means that the resolver doesn't care about (or get to pick) the version. Or in other words, a path or git dependency.

Git deps don't need precise versions to be "unresolvable": if a git dep points at HEAD, then we just fetch HEAD and that's the package we depend on. The resolver won't consider whether it should take a different revision.

On the other hand, rigid version constraints on an index package are still "resolvable" in this sense, because their presence affects the resolver's decisions.

Reply:
I see. Maybe it would be clearer to mention it explicitly ("git ref or path"), because "version" is overloaded quite a bit in our context (I initially though version would only correspond to version bounds, which is why "unresolvable version" sounded a bit paradoxical).